### DISCUSSION TITLE
Form models not updating when using browser's autocomplete feature

### DISCUSSION THREAD

**Discussion by zoltiecodes** [livewire/livewire#3070] [7. Bugs (archived)]
**Describe the bug**
Form model values are not updating when using the browser's built-in autocomplete feature. So this means when you focus a field and click on a browser-suggested value like your first name.

**To Reproduce**
1. In a component create an input field with a name attribute set to "name" and assign a model to it
2. Click the input field in the browser to focus it
3. Choose an option from the browser's autocomplete list
4. Livewire re-renders the component with an empty model value

**Expected behavior**
Livewire should pick up the change in the input field and update the model value like when you type in the field.

**Screenshots**
https://youtu.be/kVN7aQ3-_c8

**Desktop (please complete the following information):**
 - Tested in Chrome and Safari

**Additional context**
There is a workaround for this:
```
<input type="text" wire:change="$set('data.name', $event.target.value)" wire:model="data.name">
```
Listen for the change event and use `$set` to set the value. This will cause multiple XHR requests however...

**Comment by indigoram89old** [comment]
+1

**Comment by devcircus** [comment]
Other frontend frameworks have this issue as well. From past experience, you may be able to force the browser to emit a change event on autocomplete, simply by adding a "name" property to that input. 

It's more of a browser issue than a framework issue. Search the react and vue repositories for other possible workarounds.

**Comment by mihaiphilip** [comment]
Same here

**Comment by tanthammar** [comment]
+1

**Comment by chris-mackie** [comment]
+1

**Comment by calebporzio** [comment]
Ok, I recreated the issue with the following markup in a Livewire component: (I couldn't recreate it with only one or two input elements)

```html
<form action="#">
    <input type="text" name="name" wire:model="foo.name" autocomplete>
    <input type="text" name="address" wire:model="foo.address" autocomplete>
    <input type="text" name="city" wire:model="foo.city" autocomplete>
    <input type="text" name="state" wire:="foo.state" autocomplete>

    <span>@json($foo)</span>
</form>
```

It appears this issue is longstanding. Here's a stack overflow from 2013 explaining the problem and the grim outlook on solutions:

![image](https://user-images.githubusercontent.com/3670578/81626279-37acc300-93c9-11ea-9fba-c180716da79c.png)
Link: https://stackoverflow.com/questions/11708092/detecting-browser-autofill

I dug around Vue's codebase and spotted the following:
![image](https://user-images.githubusercontent.com/3670578/81626306-498e6600-93c9-11ea-89e8-ca338e2d47fe.png)
Link: https://github.com/vuejs/vue/blob/52719ccab8fccffbdf497b96d3731dc86f04c1ce/src/platforms/web/runtime/directives/model.js#L35-L50

I tried the same snippet in a codepen using Vue and it worked. It looks like Vue is forwarding the "change" event to be an "input" event to fix this bug.

Unfortunately, this is not so easy for Livewire because we would end up sending many more requests than necessary. A "change" event gets fired when you blur a text field, so not only would a request fire as the user types, but another one would fire when they tab out.

It's possible that's the only way, but because the stack overflow on that is pretty old, I'm wondering if there are any other novel solutions.

One fix for this whole problem is setting `autocomplete="off"` on your form or input tags to disable autocomplete. Assuming that's a bad UX for you, we'll have to explore other options.

Open to thoughts and ideas. Thanks.

**Comment by calebporzio** [comment]
Update: I don't think it's the browser's fault anymore.

I'm currently investigating the way Livewire handles debouncing inputs at the component level - I think this is the problem, which is why when you select auto-complete only the last input is "synced".

Because the denounce is being used which means the last one will go through and the previous won't.

I'm now more hopeful we'll be able to solve this in a clean way.

More soon.

**Comment by mihaiphilip** [comment]
Hi,

I have this issue within a login form.
When I select an auto-suggested user, only the password is synced.
For the email field to sync, I use the zsoltgyure's solution: `wire:change="$set('email', $event.target.value)"`.

Hope this detail helps :)

**Comment by georgeboot** [comment]
A workaround we use is:

```blade
<input
    {{-- wire:model="email" --}}

    x-data
    @keyup="debounce(() => @this.set('email', $el.value), 300)"
    value="{{ $email }}"
    
    type="email"
    autocomplete="email"
/>
```

**Comment by calebporzio** [comment]
Ok, the fix should be released in v1.1.0

Closing this.

**Comment by G3z** [comment]
I still have this problem with 1.1.0
<img width="791" alt="Schermata 2020-05-18 alle 10 32 59" src="https://user-images.githubusercontent.com/875086/82191538-f51c4680-98f2-11ea-9af6-31652e162271.png">

https://youtu.be/PD2qFkVhyno

it works great when typing normally

**Comment by georgeboot** [comment]
@calebporzio I'm pretty sure this is still an issue on (at least) iOS Safari 13.

**Comment by georgeboot** [comment]
@calebporzio mind taking a look at this again? I've tried it myself but can't get it to work.

I (and probably many others) would really love to use LiveWire forms for account signups and stuff, but currently it's breaking on iOS.

To reproduce, simly create a new project using https://github.com/laravel-frontend-presets/tall.

Go to the register form, and have your phone autocomplete name, email and password.

Even the above mentioned workaround (`wire:change="$set('field', $event.target.value)"`) doesnt work.

**Comment by stefro** [comment]
I'm still having these issues. They seemed to be solved in 1.3.0, but were introduced again in 1.3.1 😫

**Comment by totov** [comment]
Think I'm experiencing a similar/related issue. I have an email field and when the browser autocompletes it with a value, it fails validation due to being empty (when it's not). 

Example of this behaviour [here](https://laravelplayground.com/#/snippets/1c3ee698-6bac-4ce2-adf7-cc00ebdf77c8), but only appears to happen in Safari on macOS (Chrome and Safari on iOS seems fine).

If I turn off autocomplete using `autocomplete="off"`, then browse away from the page and return using the back button in the browser the field is emptied. This only happens when autocomplete is on the field or on the form as a whole.

**Comment by totov** [comment]
This issue was fixed in 2.2.8 https://github.com/livewire/livewire/issues/1519#issuecomment-700771163

**Comment by bureauvk** [comment]
Still experiencing this issue with Safari in macOs and iOs. Any others? 
Using version 2.3

**Comment by FranciscoDumont** [comment]
I'm using jQuey autocomplete and I've solve it adding this at the end of my blade file:

```javascript
    <script>
        document.addEventListener('livewire:load', function () {
            var availableProducts = @this.availableProducts;
            $( "#product-name-input" ).autocomplete({
                source: availableProducts,
                select: function (event, ui){
                    @this.name = ui.item.value // This is the important part
                }
            });
        })
    </script>
```

**Comment by adityar15** [comment]
@calebporzio I am using the latest Livewire version 2.3 and still having the same issue. Is there any way to fix it?

**Comment by adityar15** [comment]
> @calebporzio I am using the latest Livewire version 2.3 and still having the same issue. Is there any way to fix it?

I tried with name property to all input tags. It works well on Chrome. For the rest of the other browsers it doesn't work.

**Comment by calebporzio** [comment]
@adityar15 , can you please put together a simple code sample for others to easily reproduce the exact issue you're experiencing?

**Comment by adityar15** [comment]
Sure

```

<div class='container-fluid'>
    @section('title', 'Customer Registration')
    <h4 class="text-center fw-bold mt-5 mb-4 p-2">Registration</h4>
    @if($success)
    <div class="alert alert-success py-1 px-3 my-4">{{$success}}</div>
    @endif

    @if($error)
    <br>
    <div class="alert alert-danger py-1 px-3 my-4">{{$error}}</div>
    @endif
    <form class='register-form' wire:submit.prevent="register">

        <div class="form-group">
            <label>First Name*</label>
            <input type="text" name="firstname" class="form-control" aria-describedby="First Name" wire:model.lazy="firstname">
            @error('firstname') <div class="alert alert-danger py-1 px-3 my-3">{{ $message }}</div> @enderror
        </div>

        <div class="form-group">
            <label>Last Name*</label>
            <input type="text" name="lastname" class="form-control" aria-describedby="Last Name" wire:model.lazy="lastname">
            @error('lastname') <div class="alert alert-danger py-1 px-3 my-3">{{ $message }}</div> @enderror
        </div>

        <div class="form-group">
            <label>Email*</label>
            <input type="email" name="email" class="form-control" wire:model.lazy="email" aria-describedby="Email">
            @error('email') <div class="alert alert-danger py-1 px-3 my-3">{{ $message }}</div> @enderror
        </div>

      
        <div class="text-center">
            <button type="submit" class="btn bg-1 px-5 py-2">Register</button>
        </div>

    </form>

</div>

```


Whenever I open it on iPhone or Mac and try to fill in the data with autocomplete. It erases everything and makes fields blank.

**Reply by nakigami** [reply]
Hi, how did you solved it please ?

**Comment by calebporzio** [comment]
Sorry, please reduce this to the most minimal amount of code and provide the class portion as well. Be sure to remove all model references as well. Thanks

**Comment by HepplerDotNet** [comment]
@calebporzio got a weird issue with this. 

This [code](https://laravelplayground.com/#/snippets/4917bc68-5694-44b1-8a18-407824aaebd3) works in the playground. All fields update the properties and validation passes. 

But on my dev box with the same code, only the last field filled from autofill gets updated. This results in validation errors for all fields except last (email).

Dev box:
Laravel: 8.41.0
Livewire: 2.4.4
PHP: 8.0.6

Browser: Chrome 90 and Safari 14.1 on MacOS 10.15.7

**Reply by WanderingDaniel** [reply]
Have the same issue, so do let me know if you find any solution - I'll do vice-versa!

**Reply by daemonk12** [reply]
Same issue still persists with me, livewire v2.10.7. Thanks
